### PR TITLE
feat: Add 'Restart From Local Devfile' action

### PR DESCRIPTION
**Pull request by RomanNikitenko**
### What does this PR do?
Add 'Restart From Local Devfile' action

Screenshot 2023-03-06 at 13 06 47

### What issues does this PR fix?

https://github.com/eclipse/che/issues/18670

### How to test this PR?
- provide some changes to a devfile 
- use `Restart From Local devfile` action of the `Eclipse Che` status bar item to restart a workspace
- check that the changes are applied when the workspace is started

### CODE REVIEW

**Review by azatsarynnyy — APPROVED** [review]
I tested it by changing the command's ID in the devfile.yaml.
After restarting the DevWorkspace with a new command, the changes were applied to the DW.

**Comment by RomanNikitenko** [comment]
I've added a new commit to improve handling errors at `Restart Workspace From Local Devfile` action. For example, we should not restart a workspace if there is an error at the updating devfile step.

File: code/extensions/che-remote/package.nls.json [hunk 1]
```diff
@@ -4,5 +4,6 @@
 	"openDocumentationCommand": "Eclipse Che: Open Documentation",
 	"openDashboardCommand": "Eclipse Che: Open Dashboard",
 	"stopWorkspaceCommand": "Eclipse Che: Stop Workspace",
-	"restartWorkspaceCommand": "Eclipse Che: Restart Workspace"
+	"restartWorkspaceCommand": "Eclipse Che: Restart Workspace",
+	"restartWorkspaceFromLocalDevfileCommand": "Eclipse Che: Restart Workspace From Local Devfile"
```
Comment (rkratky) [inline code/extensions/che-remote/package.nls.json:8]: Ideally, the preposition would not be capitalized:

"Eclipse Che: Restart Workspace from Local Devfile"

File: code/extensions/che-remote/src/extension.ts [hunk 2]
```diff
@@ -46,11 +53,60 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
...
+
+  const currentDevfile = await devfileService.get()
+  const projects = currentDevfile.projects || [];
+
+  const newContent = await devWorkspaceGenerator.generateDevfileContext({ devfilePath, editorEntry: DEFAULT_EDITOR_ENTRY, projects: [] }, axios.default);
+  if (newContent) {
+    newContent.devWorkspace.spec!.template!.projects = projects;
+    await devfileService.updateDevfile(newContent.devWorkspace.spec?.template);
+  } else {
+    throw new Error('An error occurred while performing generation a new devfile context');
```
Comment (rkratky) [inline code/extensions/che-remote/src/extension.ts:110]: ->

An error occurred while generating new devfile context

File: code/extensions/che-remote/src/extension.ts [hunk 3]
```diff
@@ -46,11 +53,60 @@ export async function activate(context: vscode.ExtensionContext): Promise<void>
...
+    devfileExists = await fs.pathExists(devfilePath);
+  }
+
+  if (!devfileExists) {
+    devfilePath = await vscode.window.showInputBox({ title: 'Path to the devfile', value: `${projectPath}/` });
+    devfileExists = devfilePath ? await fs.pathExists(devfilePath) : false;
+  }
+
+  if (!devfileExists) {
+    throw new Error(`The devfile was not found by path: ${devfilePath}`);
```
Comment (rkratky) [inline code/extensions/che-remote/src/extension.ts:99]: ->

The devfile was not found at:

**Comment by RomanNikitenko** [comment]
@rkratky 
thank you, Robert, for the review!
I've applied your suggestions within another PR: https://github.com/che-incubator/che-code/pull/189

**Pull request merged**